Addition/clang 12

InfRandomness opened the pull request:
This pull requests adds clang-12

Commit:
Add clang-12

Commit:
Add 09-rpath to fix libgcc related issues

Commit:
Add clang-12.p5m

Commit:
Remove scan-view

Remove scan-view from the manifest to avoid having troubles with the dependencies

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

Commit:
Modify path of files to patch

Commit:
Replace stdlib.h by cstdlib

This change is due to stdlib.h being deprecated

Commit:
Modify positions of lines to patch

Commit:
Merge branch 'addition/clang-12' of github:InfRandomness/oi-userland into addition/clang-12

ghost commented:
This closes #6751

Commit:
Modify dependencies

AndWac commented on components/developer/clang-12/Makefile:
```diff
@@ -0,0 +1,118 @@
+
+COMPONENT_VERSION=	12.0.0
+COMPONENT_REVISION=	1
+COMPONENT_MAJOR_VERSION= \
```
You don't need to set COMPONENT_REVISION if we never have released the version before. It's an artificial discriminator to be able to change our version if the version of project hasn't changed.

InfRandomness replied:
oh, alright

Commit:
Remove COMPONENT_REVISION

COMPONENT_REVISION is not useful for a first release in oi-userland

AndWac commented on components/developer/clang-12/Makefile:
```diff
@@ -0,0 +1,117 @@
+
+COMPONENT_REVISION=	1
+COMPONENT_MAJOR_VERSION= \
+	$(shell echo $(COMPONENT_VERSION) | $(GSED) -e 's/\([0-9]\+\.[0-9]\+\).*/\1/')
```
You seem to have misunderstood me: We need COMPONENT_VERSION but not necessarily the COMPONENT_REVISION!
COMPONENT_REVISION is the part of the full version number that is under our control and only needed to be set (or incremented) when we make local changes. Thus, when we introduce a new component or update to a newer version of an existing component, we don't need to set COMPONENT_REVISION.

InfRandomness replied:
Oh... sorry, I just woke up...

AndWac commented:
You'll need to add manifest/sample-manifest.p5m and pkg5 files.
Did you check the other patches we have for older clang version aren't applicable or needed by this version, too?

ghost commented:
> You'll need to add manifest/sample-manifest.p5m and pkg5 files.

Alright.

> Did you check the other patches we have for older clang version aren't applicable or needed by this version, too?

I have applied the straight minimum to be able to compile and run a simple hello world program made in C but did not go any further; might need to

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

Commit:
Create Makefile

Revert local changes to let the CC and CXX environment variables indicate the compiler to use

Commit:
Merge pull request #1 from InfRandomness/addition/clang-12-1

Fix Makefile

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

Commit:
Add sample-manifest.p5m

ghost commented:
I am having a bunch of these while trying to run "gmake publish"
I do not understand how;
Where could a 32-bit rpath could've slipped into a 64-bit binary ?


[error.txt](https://github.com/OpenIndiana/oi-userland/files/6564808/error.txt)

AndWac commented:
The actual version does not build here:
void llvm::erase_if<llvm::SmallVector<std::pair<unsigned int, llvm::MDNode*>, 2u>, llvm::IRBuilderBase::AddOrRemoveMetadataToCopy(unsigned int, llvm::MDNode*)::{lambda(std::pair<unsigned int, llvm::MDNode*> const&)#1}>(llvm::SmallVector<std::pair<unsigned int, llvm::MDNode*>, 2u>&, llvm::IRBuilderBase::AddOrRemoveMetadataToCopy(unsigned int, llvm::MDNode*)::{lambda(std::pair<unsigned int, llvm::MDNode*> const&)#1}) 0x151a    	../../lib/libLLVMXCoreCodeGen.a(XCoreLowerThreadLocal.cpp.o)
ld: fatal: relocations remain against allocatable but non-writable sections
collect2: error: ld returned 1 exit status
make[3]: *** [tools/llvm-shlib/CMakeFiles/LLVM.dir/build.make:301: lib/libLLVM-12.so] Error 1
make[3]: Leaving directory '/oibuild/oi-userland/components/developer/clang-12/build/amd64'
make[2]: *** [CMakeFiles/Makefile2:76434: tools/llvm-shlib/CMakeFiles/LLVM.dir/all] Error 2
make[2]: Leaving directory '/oibuild/oi-userland/components/developer/clang-12/build/amd64'
make[1]: *** [Makefile:156: all] Error 2
make[1]: Leaving directory '/oibuild/oi-userland/components/developer/clang-12/build/amd64'
gmake: *** [/oibuild/oi-userland/make-rules/cmake.mk:179: /oibuild/oi-userland/components/developer/clang-12/build/amd64/.built] Error 2

ghost commented:
It builds fine for me with clang-90; are you using gcc and or GNU ld ?

AndWac commented:
On 5/29/21 4:58 PM, InfRandomness wrote:
>
> It builds fine for me with clang-90; are you using gcc ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/6756#issuecomment-850847015>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4N43HJ6IL75Z63VKQKDTQD6IXANCNFSM45A2TBRA>.
>
I am using whatever is configured in the Makefile by using "gmake
sample-manifest".

ghost commented:
Oh; hm, I do not have those errors, that's odd

ghost commented:
I have re made all of that on a clean OpenIndiana environment in a vmware virtual machine and I can't reproduce what you're having

AndWac commented:
On 5/29/21 6:11 PM, InfRandomness wrote:
>
> I have re made all of that on a clean OpenIndiana environment in a
> vmware virtual machine and I can't reproduce what you're having
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/6756#issuecomment-850857945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4N77CHESPLGSUG4VP6DTQEG3DANCNFSM45A2TBRA>.
>
I have the same result on my 2nd build server.

ghost commented:
what do they have in common ? can you give me more information about your build environment so that I can try to reproduce it ? what steps do you run ?

AndWac commented:
On 5/29/21 9:02 PM, InfRandomness wrote:
>
> what do they have in common ? can you give me more information about
> your build environment so that I can try to reproduce it ? what steps
> do you run ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/6756#issuecomment-850883704>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4N73KBZABBO6TPSAUILTQE24DANCNFSM45A2TBRA>.
>
They are both up-to-date as of today. I just run "gmake sample-manifest"
in the clang-12 directory.

AndWac commented:
On 5/29/21 9:02 PM, InfRandomness wrote:
>
> what do they have in common ? can you give me more information about
> your build environment so that I can try to reproduce it ? what steps
> do you run ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/6756#issuecomment-850883704>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4N73KBZABBO6TPSAUILTQE24DANCNFSM45A2TBRA>.
>
You wrote that you are using clang to build it. How so? There is nothing
in the Makefile that make it use clang as far as I can see. There even
no dependency on any clang version in the Makefile.

ghost commented:
> You wrote that you are using clang to build it. How so? There is nothing in the Makefile that make it use clang as far as I can see. There even no dependency on any clang version in the Makefile.

I have it modified locally, I used to have clang and clang++ harcoded but I reverted those changes back https://github.com/OpenIndiana/oi-userland/pull/6756/commits/c47a09007fae852f5e264d54009adb461f9768d2;
I used clang-90, using GCC would make me OOM with a 16gb of ram virtual machine

AndWac commented on components/developer/clang-12/Makefile:
```diff
@@ -0,0 +1,115 @@
+REQUIRED_PACKAGES += system/library
+REQUIRED_PACKAGES += system/library/math
+# Auto-generated dependencies
+REQUIRED_PACKAGES += $(GCC_RUNTIME_PKG)
```
You need to cleanup here. The REQUIRED_PACKAGES are listed twice.

Commit:
Remove duplicate REQUIRED_PACKAGES

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

Commit:
Merge branch 'OpenIndiana:oi/hipster' into addition/clang-12

ghost commented:
~~I'm still not sure where this could be coming from, OOM'd again while trying to build with GCC..
alright, so, I actually was able to reproduce this issue after a while : 
https://gist.github.com/InfRandomness/9c0139aaf9d1836ebf82fd3a4ab6b4be
I will check the available patches, to see if one already resolves this issue~~
After having looked around the error for awhile, clang-90's patch "07" had similarities with the thing causing an issue, applying it fixed the issues @AndWac

Commit:
Add shlib patch

This patches solves error(s) encountered while building clang with GCC
